File: src/commands/warReport.ts
import { Message, MessageEmbed } from 'discord.js';
import fs from 'fs';
import path from 'path';
import request from 'request';
import serverConfig from '../mongodb/serverConfig';

module.exports = {
  name: 'warReport',
  description: 'Displays War Report for the Map/Hex',
  async execute(msg: Message, args: string[]) {
    if (args[0]) {
      serverConfig.findOne({ guildID: msg.guildId }, (err: any, data: any) => {
        if (data) {
          fs.readFile(
            path.resolve(
              __dirname,
              `../cache/warReport/${args[0]}${data.shardName}.json`
            ),
            'utf-8',
            async (err, fileData) => {
              if (fileData) {
                const cache = JSON.parse(fileData);
                await apiRequest(cache, data);
              } else if (!fileData) {
                const cache = {
                  version: '0',
                };
                await apiRequest(cache, data);
              } else {
                console.log(err);
                msg.reply('An error has occured!');
              }
            }
          );
        } else if (!data) {
          msg.reply({
            content:
              'Shard setting missing, please run the command `War!setShard {shard1 | shard2}` to fix this issue!',
          });
        }
      });
    } else {
      msg.reply({ content: 'Please specify the Map/Hex for the War Report!' });
    }

    async function apiRequest(cache: any, mongoDB: any) {
      const warReportURL = {
        url: `${mongoDB.shard}/api/worldconquest/warReport/${args[0]}`,
        headers: { 'If-None-Match': `"${cache.version}"` },
      };

      try {
        request(warReportURL, async function (error, response, body) {
          switch (response.statusCode) {
            case 404:
              msg.reply({ content: 'War Report for that map wasnt found!' });
              break;
            case 200:
              console.log(
                `Api request for: WarReport, Response Code: ${response.statusCode}, File: Not Up-to Date!`
              );

              const data = JSON.parse(body);

              serverConfig.findOne(
                { guildID: msg.guildId },
                (err: any, mongoDB: any) => {
                  fs.writeFile(
                    path.resolve(
                      __dirname,
                      `../cache/warReport/${args[0]}${mongoDB.shardName}.json`
                    ),
                    JSON.stringify(data, null, 2),
                    'utf-8',
                    (err) => {
                      if (err) {
                        console.log(err);
                      } else {
                        console.log(`File: Updated!`);
                      }
                    }
                  );
                }
              );

              const warReport = new MessageEmbed()
                .setFields(
                  {
                    name: 'Total Enlistments',
                    value: `${data.totalEnlistments}`,
                  },
                  {
                    name: 'Colonial Casualties',
                    value: `${data.colonialCasualties}`,
                  },
                  {
                    name: 'Warden Casualties',
                    value: `${data.wardenCasualties}`,
                  },
                  {
                    name: 'Day Of War',
                    value: `${data.dayOfWar}`,
                  }
                )
                .setFooter('Requested at')
                .setTimestamp(new Date());
              try {
                await msg.reply({ embeds: [warReport] });
              } catch (err) {
                console.log(err);
                msg.author
                  .send(
                    'Please enable all needed permisions. Or wait for an issue to be fixed. Support server: https://discord.gg/9wzppSgXdQ'
                  )
                  .catch((err) => {
                    console.log(err);
                  });
              }
              break;
            case 304:
              console.log(
                `Api Request for: WarReport, Response Code: ${response.statusCode}, File: Up-to Date!`
              );

              const warReportCached = new MessageEmbed()
                .setFields(
                  {
                    name: 'Total Enlistments',
                    value: `${cache.totalEnlistments}`,
                  },
                  {
                    name: 'Colonial Casualties',
                    value: `${cache.colonialCasualties}`,
                  },
                  {
                    name: 'Warden Casualties',
                    value: `${cache.wardenCasualties}`,
                  },
                  {
                    name: 'Day Of War',
                    value: `${cache.dayOfWar}`,
                  }
                )
                .setFooter('Requested at')
                .setTimestamp(new Date());
              try {
                await msg.reply({ embeds: [warReportCached] });
              } catch (err) {
                console.log(err);
                msg.author
                  .send(
                    'Please enable all needed permisions. Or wait for an issue to be fixed. Support server: https://discord.gg/9wzppSgXdQ'
                  )
                  .catch((err) => {
                    console.log(err);
                  });
              }
              break;
          }
        });
      } catch (err) {
        console.log(err);
        msg.author
          .send(
            'Please enable all needed permisions. Or wait for an issue to be fixed. Support server: https://discord.gg/9wzppSgXdQ'
          )
          .catch((err) => {
            console.log(err);
          });
      }
    }
  },
};

File: src/commands/war_report.rs
use poise::serenity_prelude as serenity;
use reqwest::StatusCode;

use crate::commands::common::{
    autocomplete_map, defer_for, guild_settings, placeholder_submitted, unreachable_message,
};
use crate::utils::api_definitions::foxhole::WarReport;
use crate::utils::cache::{load_war_report, save_war_report};
use crate::utils::http;
use crate::utils::regions::display_name;
use crate::{Context, Error};

/// Responds with the war report for a single region.
#[poise::command(slash_command, guild_only)]
pub async fn war_report(
    ctx: Context<'_>,
    #[description = "Name of the region you want a report for."]
    #[autocomplete = "autocomplete_map"]
    map_name: String,
) -> Result<(), Error> {
    let Some(guild) = guild_settings(ctx).await? else {
        return Ok(());
    };

    defer_for(ctx, &guild).await?;

    if placeholder_submitted(ctx, &map_name).await? {
        return Ok(());
    }

    let cached = load_war_report(&map_name, &guild.shard_name).await;

    // Not `?`. A shard that has gone offline since it was set up fails here at
    // the transport layer, and propagating that reaches the user as "Something
    // went wrong running that command" — true, unactionable, and identical to
    // the message for a genuine bug.
    let response = match http::client()
        .get(format!(
            "{}/worldconquest/warReport/{map_name}",
            guild.shard
        ))
        .header(
            "If-None-Match",
            format!("\"{}\"", cached.as_ref().map_or(0, |r| r.version)),
        )
        .send()
        .await
    {
        Ok(response) => response,
        Err(err) => {
            log::warn!("could not reach shard {} for a war report: {err}", guild.shard_name);
            ctx.say(unreachable_message(&guild.shard_name)).await?;
            return Ok(());
        }
    };

    let report = match response.status() {
        StatusCode::NOT_MODIFIED => match cached {
            Some(report) => report,
            // Our cache went away between the read and the request; refetch
            // unconditionally rather than unwrapping a None.
            None => match refetch(&guild.shard, &map_name).await {
                Ok(report) => report,
                Err(err) => {
                    log::warn!("could not refetch the war report for {map_name}: {err}");
                    ctx.say(unreachable_message(&guild.shard_name)).await?;
                    return Ok(());
                }
            },
        },
        StatusCode::OK => {
            let report = response.json::<WarReport>().await?;
            save_war_report(&report, &map_name, &guild.shard_name).await;
            report
        }
        status => match cached {
            Some(report) => {
                log::warn!("the Foxhole API returned {status}, serving the cached war report");
                report
            }
            None => {
                ctx.say(format!(
                    "The Foxhole API returned `{status}` for that region. Try again shortly."
                ))
                .await?;
                return Ok(());
            }
        },
    };

    let embed = serenity::CreateEmbed::new()
        .color((0, 0, 0))
        .title(display_name(&map_name))
        .field("Total Enlistments", report.total_enlistments.to_string(), false)
        .field(
            "Colonial Casualties",
            report.colonial_casualties.to_string(),
            false,
        )
        .field(
            "Warden Casualties",
            report.warden_casualties.to_string(),
            false,
        )
        .field("Day Of War", report.day_of_war.to_string(), false)
        .footer(serenity::CreateEmbedFooter::new("Requested at"))
        .timestamp(serenity::Timestamp::now());

    ctx.send(poise::CreateReply::default().embed(embed)).await?;

    Ok(())
}

async fn refetch(api_url: &str, map_name: &str) -> Result<WarReport, Error> {
    let report = http::client()
        .get(format!("{api_url}/worldconquest/warReport/{map_name}"))
        .send()
        .await?
        .error_for_status()?
        .json::<WarReport>()
        .await?;

    Ok(report)
}
